scheds/rust/scx_tickless/src/main.rs

```diff
@@ -0,0 +1,349 @@
+use std::sync::atomic::Ordering;
+use std::sync::Arc;
+use std::time::Duration;
+
+use affinity::set_thread_affinity;
+use anyhow::bail;
+use anyhow::Context;
+use anyhow::Result;
+use clap::Parser;
+use crossbeam::channel::RecvTimeoutError;
+use libbpf_rs::skel::OpenSkel;
+use libbpf_rs::skel::Skel;
+use libbpf_rs::skel::SkelBuilder;
+use libbpf_rs::OpenObject;
+use libbpf_rs::ProgramInput;
+use log::warn;
+use log::{debug, info};
+use scx_stats::prelude::*;
+use scx_utils::build_id;
+use scx_utils::import_enums;
+use scx_utils::scx_enums;
+use scx_utils::scx_ops_attach;
+use scx_utils::scx_ops_load;
+use scx_utils::scx_ops_open;
+use scx_utils::set_rlimit_infinity;
+use scx_utils::uei_exited;
+use scx_utils::uei_report;
+use scx_utils::Cpumask;
+use scx_utils::UserExitInfo;
+use scx_utils::NR_CPU_IDS;
+use stats::Metrics;
+
+const SCHEDULER_NAME: &'static str = "scx_tickless";
+
+#[derive(Debug, Parser)]
+struct Opts {
+    /// Exit debug dump buffer length. 0 indicates default.
+    #[clap(long, default_value = "0")]
+    exit_dump_len: u32,
+
+    /// Define the set of CPUs, represented as a bitmask in hex (e.g., 0xff), dedicated to process
+    /// scheduling events.
+    #[clap(short = 'm', long, default_value = "0x1")]
+    primary_domain: String,
+
+    /// Maximum scheduling slice duration in microseconds (applied only when multiple tasks are
+    /// contending the same CPU).
+    #[clap(short = 's', long, default_value = "20000")]
+    slice_us: u64,
+
+    /// Frequency of the tick triggered on the scheduling CPUs to check for task time slice
+    /// expiration (0 == CONFIG_HZ).
+    ///
+    /// A higher frequency can increase the overall system responsiveness but it can also introduce
+    /// more scheduling overhead and load on the primary CPUs.
+    #[clap(short = 'f', long, default_value = "0")]
+    frequency: u64,
+
+    /// Try to keep tasks running on the same CPU
+    ///
+    /// This can help to improve cache locality at the cost of introducing some extra overhead in
+    /// the scheduler (and increase the load on the primary CPUs).
+    #[clap(short = 'p', long, action = clap::ArgAction::SetTrue)]
+    prefer_same_cpu: bool,
+
+    /// Disable SMT topology awareness.
+    #[clap(short = 'n', long, action = clap::ArgAction::SetTrue)]
+    nosmt: bool,
+
+    /// Enable stats monitoring with the specified interval.
+    #[clap(long)]
+    stats: Option<f64>,
+
+    /// Run in stats monitoring mode with the specified interval. Scheduler
+    /// is not launched.
+    #[clap(long)]
+    monitor: Option<f64>,
+
+    /// Enable verbose output, including libbpf details.
+    #[clap(short = 'v', long, action = clap::ArgAction::SetTrue)]
+    verbose: bool,
+
+    /// Print scheduler version and exit.
+    #[clap(short = 'V', long, action = clap::ArgAction::SetTrue)]
+    version: bool,
+
+    /// Show descriptions for statistics.
+    #[clap(long)]
+    help_stats: bool,
+}
+
+fn is_smt_active() -> bool {
```

Comment:
As a later cleanup this should be moved into `scx_utils` so it doesn't have to be reimplemented so many places.

Reply:
yes, good point!